content/articles/lru-cache-implementation-in-javascript/index.md

```diff
@@ -0,0 +1,140 @@
+
+[Hash Table & Double Linked List Diagram](/ddl_hash.png)
+
+To add black to our list, we will remove the rear element blue from our list. Blue in this case is our least recently accessed item. Update the rear element to point to the previous element white. In our hash table, we will remove the value(address) that is associated with our key blue. Therefore, the value for our key blue will be null. A new node will is created where we put the black. The previous front gets updated to point to the node that has black. Black will also have an address associated with it as well.
+
+The shifting, in this case, is done in 0(1).
+
+Before adding black, we need to search first and see if it is available in the cache. To search for black, we go to the hash table. In the Hash table, key black will not have a corresponding value attached to it. This means that it is not present in our cache. This allows us to find out whether an item is in the cache with just O(1) time complexity.
+
+### Implementing LRU cache in JavaScript.
+By visiting the link below, there is a leet code question to help us in the implementation of the LRU Cache. [This](https://leetcode.com/problems/lru-cache/) code implementation below can guide us in the implementation.
+
+Below are the steps we will use to implement the LRU Cache class
+
+1. We first initialize the LRU cache with a positive capacity.
+
+```javascript
+var LRUCache = function (capacity) {
+    this.capacity = capacity;
+    this.map = new Map(); //this stores the entire array
+
+    //this is boundaries for double Linked List
+    this.head = {};
+    this.tail = {};
+    this.head.next = this.tail; //initialize your double Linked List
+    this.tail.prev = this.head;
+};
+```
+2.  Get operation. This will return the value of the Key if it exists. Else, it returns -1.
+
+```javascript
+LRUCache.prototype.get = function (key) {
+    if (this.map.has(key)) {
+        //remove elem from current position
+        let c = this.map.get(key);
+        c.prev.next = c.next;
+        c.next.prev = c.prev;
+
+        this.tail.prev.next = c; //insert it after last element. Element before tail
+        c.prev = this.tail.prev; //update c.prev and next pointer
+        c.next = this.tail;
+        this.tail.prev = c; //update last element as tail
+        return c.value;
+    } else {
+        return -1; //element does not exist
+    }
+
+};
+```
+
+3.  Put operation. The put operation will update the value of the key if it is found. If found, add the `Key` and the value pair to the cache. If the number of keys has exceeded the initialized capacity of the cache, evict the least recently accessed item.
+
+```javascript
+LRUCache.prototype.put = function (key, value) {
+
+    if (this.get(key) !== -1) { // if key does not exist, update last element value
+        this.tail.prev.value = value;
+    } else {
+        //need to check if map size is at capacity
+        if (this.map.size === this.capacity) {

```

Comment:
Provide instructions on how to run this code. Also explain what to expect after running the code.